src/apps/activity_assignments/service.py

```diff
@@ -225,6 +297,99 @@ async def _get_assignments_entities(
+        )
+        return assignment_id
+
+    async def unassign_many(
+        self, applet_id: uuid.UUID, assignments_unassign: list[ActivityAssignmentCreate]
+    ) -> list[ActivityAssignment]:
+        """
+        Unassigns multiple activity assignments by marking them as deleted.
+
+        This method takes a list of assignment requests, retrieves the corresponding
+        assignment entities from the database, marks them as deleted, and then returns
+        the updated assignments.
+
+        Parameters:
+        -----------
+        applet_id : uuid.UUID
+            The ID of the applet for which the assignments are being unassigned.
+
+        assignments_unassign : list[ActivityAssignmentCreate]
+            A list of assignment creation objects that specify which assignments to unassign.
+            Each object contains details such as `activity_id`, `activity_flow_id`,
+            `respondent_subject_id`, and `target_subject_id`.
+
+        Returns:
+        --------
+        list[ActivityAssignment]
+            A list of `ActivityAssignment` objects that have been marked as deleted.
+
+        Process:
+        --------
+        1. Retrieves all relevant entities (activities, flows, subjects) using
+        `_get_assignments_entities`.
+        2. Validates each assignment and retrieves the corresponding assignment ID
+        using `_get_assignment_by_activity_or_flow_and_subject_id`.
+        3. Creates a schema for each assignment, marking it as deleted.
+        4. Unassigns the assignments in the database using `ActivityAssigmentCRUD.unassign_many`.
+        5. Optionally, sends emails based on the respondent activities (not yet implemented).
+        6. Returns a list of the updated `ActivityAssignment` objects.
+
+        """
+        entities = await self._get_assignments_entities(applet_id, assignments_unassign)
+
+        respondent_activities: dict[uuid.UUID, list[str]] = defaultdict(list)
+        schemas = []
+        for assignment in assignments_unassign:
+            activity_or_flow_name: str = self._validate_assignment_and_get_activity_or_flow_name(assignment, entities)
+
+            id_assignment = await self._get_assignment_by_activity_or_flow_and_subject_id(assignment)
+            schema = ActivityAssigmentSchema(
+                id=id_assignment,
+                activity_id=assignment.activity_id,
+                activity_flow_id=assignment.activity_flow_id,
+                respondent_subject_id=assignment.respondent_subject_id,
+                target_subject_id=assignment.target_subject_id,
+                is_deleted=True,
+            )
+
+            if entities.respondent_subjects[assignment.respondent_subject_id].user_id:
+                respondent_activities[assignment.respondent_subject_id].append(activity_or_flow_name)
+
+            schemas.append(schema)
+
+        assignment_schemas: list[ActivityAssigmentSchema] = await ActivityAssigmentCRUD(self.session).unassign_many(
```

Comment:
we don't need thoses validations here, we are not creating assignments we are updating assignments that already exists

so, search for assignments that for "activity or flow" and "respondent and target", then update is_deleted field, this logic should be inside "unassign_many" method

Reply:
+1 – agreed with @rcmerlo this can be a lot simpler. We shouldn't be calling `_get_assignments_entities` – it's costly and not necessary for this action.

Reply:
You mean the ActivityAssigmentCRUD.unassign_many method correct?

Reply:
I coded the search id_assignment search in this script (service.unassign_many )instead of ActivityAssigmentCRUD.unassign_many trying to keep the same structure as service.create_many. Which searches and obtains all the schemas and then calls ActivityAssigmentCRUD
(This section)
`
        for assignment in assignments_unassign:
            id_assignment = await self._get_assignment_by_activity_or_flow_and_subject_id(assignment)
            schema = ActivityAssigmentSchema(
                id=id_assignment,
                activity_id=assignment.activity_id,
                activity_flow_id=assignment.activity_flow_id,
                respondent_subject_id=assignment.respondent_subject_id,
                target_subject_id=assignment.target_subject_id,
                is_deleted=True,
            )
            schemas.append(schema)
`

You are correct about the:

         assignment_schemas: list[ActivityAssigmentSchema] = await ...
 it doesn't requires validation.  Changes :D

Reply:
Validations are now deleted and instead of using using _update I update all the entries in the DataBase at the same time using 
                    tuple_(
                        ActivityAssigmentSchema.activity_id,
                        ActivityAssigmentSchema.respondent_subject_id,
                        ActivityAssigmentSchema.target_subject_id,
                    )
                    
 as an id to identify the correct assignment and then replace is_delete value